packages/jss/src/PluginsRegistry.js

```diff
@@ -71,6 +71,7 @@ export default class PluginsRegistry {
     for (let i = 0; i < this.registry.onProcessStyle.length; i++) {
       // $FlowFixMe
       rule.style = this.registry.onProcessStyle[i](rule.style, rule, sheet)
+      if (!rule.style) return
```

Comment:
onProcessStyle  was supposed to always return a new style object, can we pls move this check to the plugin?

Reply:
The function rule plugin needs to return previous style value when user function returns empty

Reply:
onProcessStyle was designed similar to a [].map function

Reply:
For me it looks more like a `[].reduce`. Given the following arguments, you might want to reconsider it:

- It would make behavior inconsistent between passing a null|undefined style directly and having a plugin that produces this output. 
- Given example of the problem I provided, the style is not an `{...}`, but a function. So basically the plugin would have to invoke the function, check for null|undefined and if it is, return the function back if its null|undefined. You would expect the user function being a map of input -> style, but it behaves differently to 1. providing null|undefined as value (works) 2. what user would expect as a result of returning nothing (no style rule applied, like in 1.)
- You would have to implement checks in every plugin (even some internal fail with exception because currently they do not, for example keyframes internal).
- You would run hooks even thought it does not make any sense to run them

The last two points also have performance implications. While if it did make sense, it would be not a question of performance, but it also does not make much sense in reflection of the first two points, so why? It makes it also more error prone, because now you can do more things wrong in plugins and you can do more things wrong as consumer of such a plugin, by expecting a function plugin to resolve to value returned and not to the function provided itself, because it returned null|undefined. And this is only an example of issues I encountered in short period of time. I am pretty sure this leads to more soon.

Reply:
> The function rule plugin needs to return previous style value when user function returns empty

I don't think that's correct. It should return an empty object in my opinion.

Reply:
@HenriBeck if there was no style object before - yes it should return empty object

Reply:
@kof
> if there was no style object before - yes it should return empty object, I think though there is always a style object, because there should be an empty one by default

There can be no style object, it you pass a function as `rule.style` and the function value plugin is active.

@HenriBeck 
Then [this](https://github.com/akomm/jss/blob/fix-1159/packages/jss/src/PluginsRegistry.js#L62) should not bail out, but run call `onProcessStyle` with empty style object too.

Also note, `null|undefined` is easier checked for changes, than new empty object created on each run/update, even though it did not change and is still empty.

Reply:
I have tried to reproduce/fix it and ended up with this 
 https://github.com/cssinjs/jss/pull/1164 

I tried first to support  a falsy value directly in the core, but this resulted in so many falsy value checks in different places that I immediately regretted it and reverted to just using a default empty object in the plugin.

I think this is an edge case.

Reply:
I wonder if we should warn when onProcessStyle doesn't return an object to ensure this connsistently.

Reply:
The question preventing me from moving it to the plugin in this PR is still unaswere though.

Reply:
The question preventing me from moving it to the plugin in this PR is still unaswered though.
Btw, bailing out would have required only 1 check AFAIK. Because other plugins would not get to check, as they never get falsy value then. But nvm.

Reply:
What argument is there to continue process onProcessStyle further plugins with empty object? So that maybe a scenario is possible where plugin inserts something into this empty object? Okay. Might be useful. Then why should not be the same possible when doing exactly same thing just without function value plugin? See https://github.com/cssinjs/jss/pull/1160#discussion_r303336699

Reply:
> Btw, bailing out would have required only 1 check AFAIK. Because other plugins would not get to check, as they never get falsy value then. But nvm.

yeah and then we have a behaviour where any plugin can prevent other hooks being called by returning null, that wouldn't be easy to debug.

Reply:
>Then why should not be the same possible when doing exactly same thing just without function value plugin? See #1160 (comment)

Can you explain this pls?

Reply:
`PluginRegistry.onProcessRule` does falsy-check `rule.style`. If its true, `PluginRegistry.onProcessStyle` is invoked. 

This means, when you provide style this way:
```js
jss.createStyleSheet({a: undefined})
```
**No** `onProcessStyle` hooks will be invoked. The resulting `rule.style` is falsy.

When you provide the style this way:
```js
jss.createStyleSheet({a: () => undefined})
```
**All** `onProcessStyle` hooks will be invoked. The resulting `rule.style` is (optimistically) never falsy, as falsy must be replaced with empty object by plugins.

You would except same outcome, but the outcome is different for this two scenarios.

If you do not want to handle falsy values by those `onProcessStyle` hooks, hence the check before in `onProcessRule`, why do you want to handle falsy values (by replacing with empty object) afterwards? 

Is there any good reason doing so? I gave an example, what it could make possible: A plugin that injects something into the empty object. If we would exit early, it would not be possible. Then the final question is, why should it not be possible when the style is false on first pass through `onProcessRule` too? Or the other way around: if bailing out on falsy value is preferred way and there is no intention to inject something via plugin into empty style objects, why do we do it in `onProcessStyle` ? It then just pointlessly runs through hooks and also provides inconsistent behavior where you expect consistency.

Reply:
At least, given current fix you merged, `onProcessRule` should initialize `rule.style` to empty object, if it is falsy, instead of skipping `onProcessStyles`. Or you might reconsider to do it the other way around.

Reply:
> PluginRegistry.onProcessRule does falsy-check rule.style. If its true, PluginRegistry.onProcessStyle is invoked.

Yes the idea is that no `onProcessStyle` hook is invoked if there is no style to process. The hook is called `onProcessStyle`, sounds logical to me. It doesn't stop the call chain of that hook in the middle, it doesn't start it at all if there is no style.

Reply:
I think the main reason for the StyleRule.style to not been set to empty object by default is the emptiness check in some cases, because it's more work to check if `{}` is empty than a falsy value.

Reply:
> All onProcessStyle hooks will be invoked. The resulting rule.style is (optimistically) never falsy, as falsy must be replaced with empty object by plugins.

I agree though that in this case style is technically empty but is initialized with an empty object. We have been treeting function value/rules as an edge case so far, because in those cases, we don't know what style object will be in the future.

Reply:
Evtl. making sure that style object is never falsy is a better way for the future. cc @HenriBeck

Reply:
Another thing is that the check `if (rule.style)` in PluginsRegistry also checks it for rules which don't have a style map at all, which means that check will still have to exist.

Reply:
That check would be more explicit if it would be checking specific rule types which are supposed to have style object, but that would limit the plugins, because a plugin wouldn't be able to create a custom rule with a style prop,  whatever reason for this could be.

Reply:
> I think the main reason for the StyleRule.style to not been set to empty object by default is the emptiness check in some cases, because it's more work to check if `{}` is empty than a falsy value.

Would this in `onProcessRule`:
```js
if (!rule.style) rule.style = {};
this.onProcessStyle(rule.style, rule, sheet)
```

Not still be a falsy check?

Reply:
There are other places I believe which would go iterate over empty style object even though it's empty, but it might be not too bad, gotta check.

In any case default object should be set in the corresponding rule classes.

Reply:
Adding //@flow to the tests and then changing JssStyle type to something more specific would reveal all the places probably, because currently it's just `Object` and that's just as `any` in the latest flow versions.

Reply:
Let's open a separate issue for this investigation? Because this PR got diverged from it's original purpose and I would rather close it and start fresh.

Reply:
good idea